sycl/plugins/level_zero/pi_level_zero.hpp

```diff
@@ -217,6 +215,27 @@ struct _pi_context : _pi_object {
     // we don't need a map with device as key.
     HostMemAllocContext = new USMAllocContext(
         std::unique_ptr<SystemMemory>(new USMHostMemoryAlloc(this)));
+
+    // If we have a single device/sub-device, construction is done.
+    if (NumDevices == 1)
+      return;
+
+    // Check if we have context with subdevices of the same device (context may
+    // include root device itself as well)
+    RootDevice = Devices[0]->RootDevice ? Devices[0]->RootDevice : Devices[0];
+    for (auto &Device : Devices) {
+      if ((!Device->RootDevice && Device != RootDevice) ||
+          (Device->RootDevice && Device->RootDevice != RootDevice)) {
+        RootDevice = nullptr;
+        break;
+      }
+    }
+  }
+
+  // Utility function which returns a root-device if the context consists of
+  // sub-devices of a same device or returns the first device otherwise.
+  pi_device getFirstOrRootDevice() {
+    return RootDevice ? RootDevice : Devices[0];
```

Comment:
it is still not clear how this function is to be used. In what circumstances is it no difference if the right root device is returned or some first device only? I think we should split this (or code using this) into multiple paths.

Reply:
Idea was to use RootDevice for special case which we can support and to save old behavior for all other cases by using Devices[0].

Reply:
I understand that and agree to that. But let's not do it by adding an interface that doesn't make sense. Just where we currently use Device[0] do the right thing when we can, and continue to use Device[0] when we cannot (or add an assertion if using Device[0] absolutely not going to work based on our current knowledge).

Reply:
I see your concern, I tried to address it in the last update.